src/osd/PrimaryLogPG.cc

```diff
@@ -1529,6 +1529,44 @@ int PrimaryLogPG::do_scrub_ls(MOSDOp *m, OSDOp *osd_op)
   return r;
 }
 
+// it is pretty much the same as CEPH_OSD_OP_COPY_FROM, but it requires user to
+// specify the src osd, and uses repair_read() to read the object payload, also
+// the src oid is always identical to the dest oid (the one in OpContext)
+int PrimaryLogPG::do_repair_copy(OpContext *ctx, const OSDOp& osd_op, bufferlist::iterator& bp)
+{
+  repair_copy_arg_t arg;
+
+  vector<pg_shard_t> bad_shards;
+  try {
+    arg.decode(bp);
+  } catch (buffer::error&) {
+    dout(10) << "corrupted " << __func__ << " arg" << dendl;
+    return -EINVAL;
+  }
+  if (arg.bad_shards.empty() || !arg.what) {
+    return 0;
+  }
+  const auto& oi = ctx->obs->oi;
+  if (!ctx->copy_cb) {
+    // mark the bad shard missing so they are not chosen when repairing
+    for (const auto& bad_shard : arg.bad_shards) {
+      peer_missing[bad_shard].add(oi.soid, oi.version, eversion_t());
+    }
+    start_repair_copy(ctx, arg.what, arg.bad_shards);
+    return -EINPROGRESS;
+  } else {
+    // finish
+    if (pool.info.require_rollback()) {
```

Comment:
You unconditionally added them to the missing sets, right?  Seems like you need to unconditionally remove them?